01-indoor-floor-level-tracker/web-app/src/services/AppService.ts

```diff
@@ -1,15 +1,24 @@
 import { DataProvider } from "./DataProvider";
 import { AttributeStore } from "./AttributeStore";
-import { DeviceTracker, TrackerConfig } from "./AppModel";
+import { AuthToken, DeviceTracker, TrackerConfig } from "./AppModel";
 import { IDBuilder } from "./IDBuilder";
 import { FleetID } from "./DomainModel";
 
 // this class / interface combo passes data and functions to the service locator file
 interface AppServiceInterface {
-  setDeviceName: (deviceUID: string, name: string) => Promise<void>;
-  getDeviceTrackerData: () => Promise<DeviceTracker[]>;
-  getTrackerConfig: () => Promise<TrackerConfig>;
-  setTrackerConfig: (trackerConfig: TrackerConfig) => Promise<void>;
+  getAuthToken: () => Promise<AuthToken>;
+  checkAuthTokenValidity: (authTokenString: string) => boolean;
+  setDeviceName: (
+    authToken: AuthToken,
+    deviceUID: string,
+    name: string
+  ) => Promise<void>;
+  getDeviceTrackerData: (authToken: AuthToken) => Promise<DeviceTracker[]>;
+  getTrackerConfig: (authToken: AuthToken) => Promise<TrackerConfig>;
+  setTrackerConfig: (
+    authToken: AuthToken,
+    trackerConfig: TrackerConfig
+  ) => Promise<void>;
```

Comment:
I like the 'functional' aspect of adding an `authToken` argument to each function, but might it save some typing/reading/thinking to just have a `setAuthToken: (AuthToken) => {};` function (maybe it replaces the checkAuthTokenValidity too and just throws an error if the token is invalid?) on the `AppServiceInterface`? I'm really on the fence here. 

Remember to think from the highest-possible-level of abstraction for what makes sense there. Don't change the app layer or domain layer too precisely to suit the needs of the lower-level implementation details. If you _must_ add something to the domain or app layer, do it in a very generic way and let lower levels implement and inject details as they see fit.

I'm trying to explain a lot in a small number of words here. If this is enough detail, great, but if you want to huddle on slack let me know. Thanks for your patience and persistence here so we can make this example project exemplary.

Reply:
@CarltonHenderson just to make sure I'm understanding what you're saying: are you recommending something along the lines of this inside of `AppService.ts`?

```
interface AppServiceInterface {
  setAuthToken: (AuthToken) => {};
  // other functions...
  getDeviceTrackerData: () => Promise<DeviceTracker[]>;
}

 setAuthToken: (AuthToken) {
    return this.dataProvider.setAuthToken();
  }

  async getDeviceTrackerData() {
    const authToken = this.dataProvider.setAuthToken();
    return this.dataProvider.getDeviceTrackerData(authToken);
  }
```

Reply:
Almost... more like

```

Reply:
Or like this

```
interface AppServiceInterface {
  setAuthToken: (AuthToken) => {};
  // other functions...
  getDeviceTrackerData: () => Promise<DeviceTracker[]>;
}

 setAuthToken: (AuthToken) {
    this.authToken = this.dataProvider.setAuthToken(AuthToken);
  }

  async getDeviceTrackerData() {
    return this.dataProvider.getDeviceTrackerData(AuthToken);
  }
```

Not sure which makes more sense in this circumstance.

Actually, what probably makes most sense would be making a dataProvider that does its own auth (whether or not the notehub-js handles some of the auth + token renewal) so the AppServiceInterface doesn't need to change at all. Let me know if you want me to really put pen to paper and help figure out which one of these makes most sense.

Reply:
I hope you appreciate being given four options and freedom to choose your own adventure. Let me know if you want to pair on it. It's not that easy to figure out what will give the easiest understanding and future flexibility.

Reply:
My first attempts at making the NotehubDataProvider do its own auth is what ended up being a non-starter due to the fact that the OAuth token values weren't being preserved server-side between app refreshes.

I too would like to make the `AppServiceInterface` unchanged, but I'm just not sure it's possible. Today, I addressed your comments above and DRYed up the client side code, tomorrow, I'll try to look more deeply into these suggestions and may ping you to discuss.

Reply:
Net net after pairing with Carl, we're going back to basics: new OAuth token every API call for this app to avoid all the extra hoops this app is making us jump through to store tokens until they expire.

I'll update another app that uses a Prisma DB to store tokens there and demonstrate one way to store that data in a table and periodically update it as needed all on the server side. The client shouldn't need to know about it all